Allow sysroots and test multiple _FORTIFY_SOURCE levels

siddhesh opened the pull request:
This changeset adds the following:

- a SYSROOT option to allow testing glibc fortify headers installed in a different sysroot; it basically passes a -isysroot flag to the compiler

- Test both `-D_FORTIFY_SOURCE=1` and `-D_FORTIFY_SOURCE=2`.   I tried to auto-generate targets per fortify level but it didn't work, so for now they're hard coded.  Not a big deal though since one only needs to add one line per compiler for an additional fortification level.

siddhesh committed:
Disallow overriding _FORTIFY_SOURCE and allow SYSROOT

This is in preparation for running tests for both _FORTIFY_SOURCE=1
and _FORTIFY_SOURCE=2.  The SYSROOT variable allows tests to use
custom glibc headers so that those can be tested without installing in
the system.

siddhesh committed:
Test for multiple fortify levels

Build and run tests with -D_FORTIFY_SOURCE=1 and 2.  Also rework code
to allow easy adding of additional fortification levels in future.
See also:

https://sourceware.org/pipermail/libc-alpha/2020-November/120191.html

siddhesh committed:
Add a quiet mode by default

Use make V=1 to disable.

siddhesh commented:
Added another commit that quietens the output a bit by default; it can be overridden by passing `Q=` to make.

serge-sans-paille commented:
I thought the defacto standard was ``make V=1`` when going in verbose mode, at least that's what automake generates. It's probably better to follow that practice?

siddhesh commented:
Updated last patch to use `V=1` for verbose.

serge-sans-paille commented:
For more auto-generated stuff, you could take inspiration from

```make
define gen =
$(1)-$(2):
    echo $(2)
endef

$(foreach F,a1 a2,$(foreach G,1 2,$(eval $(call gen,$(F),$(G)))))
```

siddhesh committed:
Further auto-expansion of targets

Use Serge's user defined functions idea to expand all combinations of
compilers and fortification levels.

siddhesh commented:
There you go, fully auto-generated now :)

siddhesh committed:
Update README to mention additional coverage and targets

siddhesh commented:
Another tiny update to update `README.rst`.

serge-sans-paille commented on Makefile:
```diff
@@ -1,49 +1,65 @@
+FORTIFY_LEVELS = 1 2
 
-TARGETS=test_memcpy test_memmove test_mempcpy test_memset test_snprintf test_sprintf test_stpcpy test_strcat test_strcpy test_strncat test_strncpy test_vsnprintf test_vsprintf
+TARGETS=memcpy memmove mempcpy memset snprintf sprintf stpcpy strcat strcpy strncat strncpy vsnprintf vsprintf
```
nice!

serge-sans-paille commented on Makefile:
```diff
@@ -1,49 +1,65 @@
 
-check-target = check-$(1):$(patsubst test_%,run_%.$(1),$(TARGETS)); \
+define check-target =
+check-$(1):$(patsubst %,run_%.$(1),$(TARGETS))
```
yeah, that's much cleaner like this

serge-sans-paille commented:
Waiting for travis to complete and let's roll!

siddhesh committed:
Fix clean and dist targets for change in TARGETS

siddhesh commented:
Gah, forgot to update the `dist` and `clean` targets. Done now.
